### ISSUE TITLE
Slow Performance on AWS Lambda

### ISSUE THREAD

**Issue by dtothefp** [halt-hammerzeit/webpack-isomorphic-tools#17]
I'm rendering my React template on AWS lambda and just added npm debug module to see where the bottleneck is. Looks like it's when I `require` my react component that I am pre-compiling through Babel. I don't see this performance hit locally. I'm wondering if redefinition of `require` hooks in `webpack-isomorphic-tools` could be slowing down performance? There is a 20s lag in the logs below when the component is `require`d as well as a 13s lag when initial function is `require` inside of the isomorphic tools Promise. I can try disabling ismorphic tools and see if it helps but wanted to ask if you have any ideas of the top of your head for performance hit?

tools entry
```js
import path from 'path';
import {assign} from 'lodash';
import WebpackIsomorphicTools from 'webpack-isomorphic-tools';
import makeConfig from '../gulp/config';
import makeIsomporphicToolsConfig from '../gulp/tasks/webpack/webpack-isomorphic-tools';
import makeBootstrap from '../src/js/utils/bootstrap';
import reactor from '../src/js/modules/flux';

/**
 * Run the app with lambda or start the local dev server
 * @param {Object} data mock form data or data passed from lamda
 * @param {Boolean:undefined} isLocal true if module is run directly (ie. not required by lambda)
 * @param {Function} cb callback args => (err, html)
 * @return {undefined} use the cb
 */
export default function({data, isLocal}, cb) {
  const debug = require('./utils/run-debug')(__filename);
  const {NODE_ENV} = process.env;
  const config = makeConfig({ENV: NODE_ENV});
  const {environment, sources, utils} = config;
  const {isDev} = environment;
  const {buildDir, statsFile} = sources;
  const {logError} = utils;

  const toolsConfig = makeIsomporphicToolsConfig({statsFile: path.basename(statsFile)});
  const tools = new WebpackIsomorphicTools(toolsConfig);
  const bootstrap = makeBootstrap({
    reactor,
    data,
    env: 'server'
  });
  const statsDir = path.resolve(__dirname, '..', buildDir);

  debug('Running Isomporphic Tools');
  tools.development(isDev).server(statsDir).then(() => {
    debug('Running bootstrap#init');

    return bootstrap.init().then((props) => {
      const templatingArgs = {
        data: assign({}, data, props),
        tools,
        config: config,
        props
      };

      //do dynamic require because we don't need express to be copied for the lamda
      if (isLocal) {
        debug('Calling local server');
        const runServer = require('./dev-server');
        runServer(templatingArgs, cb);
      } else {
        debug('Calling lambda function');
        const runLambda = require('./lambda-function');
        runLambda(templatingArgs, cb);
      }
    }).catch((err) => {
      logError({err, plugin: '[app: bootstrap.inti]'});
    });
  });
}
```

debugging code
```js
import async from 'async';
import glob from 'globby';
import ReactDOM from 'react-dom/server';
import React from 'react';

export default function(config) {
  const debug = require('../utils/run-debug')(__filename);
  const {utils} = config;
  const {addbase} = utils;

  function _transform(data) {
    return function(fp, cb) {
      debug('[assemble: transform] require snippet');
      fp = addbase(fp);
      const Snippet = require(fp);
      debug('[compiling: start] ReactDOM#renderToString');
      const contents = ReactDOM.renderToString(React.createElement(Snippet, data));
      debug('[compiling: end] ReactDOM#renderToString');
      cb(null, {path: fp, contents});
    };
  }

  return function(collection) {
    debug('[assemble: loader] Initializing JSX Collection');
    collection.load = function(patterns, options, data, cb) {
      options = options || {};
      const files = glob.sync(patterns, options);

      async.map(files, _transform(data), function(err, results) {
        if (err) {
          cb(err);
        } else {
          results.forEach(function(file) {
            collection.addView(file.path, file);
          });
          debug('[assemble: loader] Finished Compiling JSX Collection');
          cb();
        }
      });
    };
  };
}
```

Lambda output
```sh
DEBUG: Wed, 04 Nov 2015 00:09:17 GMT bin/runner.js LOADING bin/runner.js => BRANCH dfp/update-environmentalization 
DEBUG: Wed, 04 Nov 2015 00:09:17 GMT bin/runner.js [lambda: init] 
DEBUG: Wed, 04 Nov 2015 00:09:17 GMT app/index.js LOADING app/index.js => BRANCH dfp/update-environmentalization 
DEBUG: Wed, 04 Nov 2015 00:09:17 GMT app/index.js Running Isomporphic Tools 
DEBUG: Wed, 04 Nov 2015 00:09:17 GMT app/index.js Running bootstrap#init 
DEBUG: Wed, 04 Nov 2015 00:09:17 GMT app/index.js Calling lambda function 
DEBUG: Wed, 04 Nov 2015 00:09:30 GMT app/lambda-function.js LOADING app/lambda-function.js => BRANCH dfp/update-environmentalization 
DEBUG: Wed, 04 Nov 2015 00:09:30 GMT lib/template-assemble.js LOADING lib/template-assemble.js => BRANCH dfp/update-environmentalization 
DEBUG: Wed, 04 Nov 2015 00:09:37 GMT lib/jsx-loader.js LOADING lib/jsx-loader.js => BRANCH dfp/update-environmentalization 
DEBUG: Wed, 04 Nov 2015 00:09:37 GMT lib/jsx-loader.js [assemble: loader] Initializing JSX Collection 
DEBUG: Wed, 04 Nov 2015 00:09:37 GMT lib/template-assemble.js Calling tempalateAssemble#renderTemp 
DEBUG: Wed, 04 Nov 2015 00:09:37 GMT lib/jsx-loader.js [assemble: transform] require snippet 
DEBUG: Wed, 04 Nov 2015 00:09:57 GMT lib/jsx-loader.js [compiling: start] ReactDOM#renderToString 
DEBUG: Wed, 04 Nov 2015 00:10:00 GMT lib/jsx-loader.js [compiling: end] ReactDOM#renderToString 
DEBUG: Wed, 04 Nov 2015 00:10:00 GMT lib/jsx-loader.js [assemble: loader] Finished Compiling JSX Collection 
DEBUG: Wed, 04 Nov 2015 00:10:00 GMT lib/template-assemble.js [rendering: start] tempalateAssemble#renderTemp 
DEBUG: Wed, 04 Nov 2015 00:10:02 GMT lib/template-assemble.js [rendering: end] tempalateAssemble#renderTemp 
DEBUG: Wed, 04 Nov 2015 00:10:02 GMT bin/runner.js [lambda: upload-success] 
[lambda]: Success uploading lambda data 
DEBUG: Wed, 04 Nov 2015 00:10:02 GMT services/s3-uploader.js LOADING services/s3-uploader.js => BRANCH dfp/update-environmentalization 
DEBUG: Wed, 04 Nov 2015 00:10:02 GMT services/s3-uploader.js UPLOADING TO MASTER: false 
DEBUG: uploaded file
Duration: 46100 ms
```

**Comment by halt-hammerzeit** [comment]
Well, when you call `require()` which is overrided it simply takes the value from `webpack-assets.json` and returns it.
In production this file is loaded and parsed a single time and then cached.
There should be no bottleneck.

You can test for a bottleneck more precisely if you do, say, 
```
console.log(new Date())
require('./asset.css')
console.log(new Date())
```

**Comment by dtothefp** [comment]
Uggg...sorry for this. I tried tons of stuff, removing the babel runtime and getting rid of babel using bluebird for promises and just using the `babel-polyfill`. The best result was getting rid of `isomorphic-tools` entirely along with all those other changes but only got it down to 32 seconds. 

What ended up being the issue is that the lambda memory had to be cranked up. I was on 128MB and went up to the max of 1536MB then it ran in 3 seconds. 

Sorry for opening an unrelated issue but maybe it will help someone else out in the future. 

Here is a gist of all of the use cases I tried

https://gist.github.com/dtothefp/0f07b05bba1edc119a6f
